zerver/lib/actions.py

```diff
@@ -1875,12 +1877,27 @@ def extract_recipients(s: Union[str, Iterable[str]]) -> List[str]:
     if not isinstance(data, list):
         raise ValueError("Invalid data type for recipients")
 
-    recipients = data
+    if all(isinstance(elem, int) for elem in data):
+        return list(set(data))
+
+    recipients = []  # type: List[Any]
+    for recipient in data:
+        recipient = recipient.strip()
+        if recipient.isdigit():
+            recipients.append(int(recipient))
+        elif recipient:
+            # Just append the stripped recipient sting
+            # if they are not an integer ID
+            recipients.append(recipient)
+
+    is_heterogenous = (not all(isinstance(recipient, int) for recipient in recipients) and
```

Comment:
I'm still not convinced we need this `is_heterogenous` complexity.  In my mind, we do this:

- split the list
- if empty, return an error
- look at first element, and based on its type, call either `validate_user_ids` or `validate_emails` with lists of strings

Then in each validation function just enforce that elements are either ints or emails, respectively.